tests/posix/common/src/pthread.c

```diff
@@ -465,6 +465,54 @@ ZTEST(pthread, test_pthread_cleanup)
 	zassert_ok(pthread_join(th, NULL));
 }
 
+static void *test_pthread_cancel_fn(void *arg)
+{
+	int oldtype = ~PTHREAD_CANCEL_DEFERRED;
+	int oldstate = ~PTHREAD_CANCEL_ENABLE;
+
+	zassert_ok(pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &oldtype));
+	zassert_equal(oldtype, PTHREAD_CANCEL_DEFERRED);
+
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldstate));
+	zassert_equal(oldstate, PTHREAD_CANCEL_ENABLE);
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	/* this will mark it pending */
+	pthread_cancel(pthread_self());
+
+	/* assume this completes well before 500ms */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &oldstate));
+
+	/* this should terminate the thread*/
+	pthread_testcancel();
+
+	/*
+	 * We use this large delay to indicate that the thread in fact did not exit via
+	 * pthread_testcancel(), as expected. I.e. it should not take > 1000 ms to join thread
+	 */
+	zassert_ok(k_msleep(1000));
+
+	return NULL;
+}
+
+ZTEST(pthread, test_pthread_testcancel)
+{
+	uint32_t now;
+	uint32_t then;
+	pthread_t th;
+
+	then = k_uptime_get_32();
+	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
+	zassert_ok(k_msleep(500));
+	zassert_ok(pthread_join(th, NULL));
+	now = k_uptime_get_32();
+
+	/* should not take > 1000 ms to join thread */
+	zassert_true(now - then <= 1000);
```

Comment:
to check if the thread is cancelled by the `pthread_testcancel()`, this should be _much_ lesser than `1000` (in fact we expect it to take just a little longer than the delay defined in L508)

Reply:
see also https://github.com/zephyrproject-rtos/zephyr/pull/67020#discussion_r1473804745